OpenApiQuery/IMaterializer.cs

```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OpenApiQuery
+{
+    public interface IMaterializer
```

Comment:
EF solves this topic by checking if the `IQueryable` is an `IAsyncEnumerable`:

https://github.com/dotnet/efcore/blob/0e7e329a24deae8d22ccf9829957dc8cb10152c5/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs#L2019
https://github.com/dotnet/efcore/blob/0e7e329a24deae8d22ccf9829957dc8cb10152c5/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs#L2815

We can do the same. Having an IMaterializer is maybe still nice for adopting libraries that do not provide yet [async enumerables](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iasyncenumerable-1?view=dotnet-plat-ext-3.1) but this way we can provide a default implementation and only need to provide extension libraries for the frameworks not having them.  

EF seems to use them already. For the MongoDB driver unfortunately no mapping exists.  http://blog.i3arnon.com/2015/12/16/async-linq-to-objects-over-mongodb/

Reply:
See below